eng/_util/cmd/sign/sign.go

```diff
@@ -0,0 +1,287 @@
+	// terminated. There are some ctx.Err() checks sprinkled into the Go code, but canceling
+	// quickly during the packaging/repackaging work in Go is not currently important: the Go work
+	// takes an insignificant amount of time compared to the signing service calls in MSBuild.
+	var ctx context.Context
+	if *timeout == 0 {
+		ctx = context.Background()
+	} else {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithDeadline(context.Background(), time.Now().Add(*timeout))
+		defer cancel()
+	}
+
+	archives, err := findArchives(ctx, *filesGlob)
+	if err != nil {
+		return err
+	}
+
+	log.Println("Signing individual files extracted from archives")
+
+	individualFilesToSign, err := flatMapSlice(archives, func(a *archive) ([]*fileToSign, error) {
+		return a.prepareEntriesToSign(ctx)
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := sign(ctx, "1-Individual", individualFilesToSign); err != nil {
+		return err
+	}
+
+	for _, a := range archives {
+		if err := a.repackSignedEntries(ctx); err != nil {
+			return err
+		}
+	}
+
+	if *notarize {
+		log.Println("Notarizing macOS archives")
+
+		filesToNotarize, err := flatMapSlice(archives, func(a *archive) ([]*fileToSign, error) {
+			return a.prepareNotarize(ctx)
+		})
+		if err != nil {
+			return err
+		}
+
+		if err := sign(ctx, "2-Notarize", filesToNotarize); err != nil {
+			return err
+		}
+
+		for _, a := range archives {
+			if err := a.unpackNotarize(ctx); err != nil {
+				return err
+			}
+		}
+	} else {
+		log.Println("Skipping notarizing macOS archives")
+	}
+
+	log.Println("Creating signature files")
+
+	signatureFiles, err := flatMapSlice(archives, func(a *archive) ([]*fileToSign, error) {
+		return a.prepareArchiveSignatures(ctx)
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := sign(ctx, "3-Sigs", signatureFiles); err != nil {
+		return err
+	}
+
+	log.Println("Copying finished files to destination")
+
+	for _, a := range archives {
+		if err := a.copyToDestination(ctx); err != nil {
+			return err
+		}
+	}
+
+	log.Println("Generating checksum files")
+
+	for _, a := range archives {
+		if err := checksum.WriteSHA256ChecksumFile(filepath.Join(*destinationDir, a.name)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func findArchives(ctx context.Context, glob string) ([]*archive, error) {
+	files, err := filepath.Glob(glob)
+	if err != nil {
+		return nil, fmt.Errorf("failed to glob files: %v", err)
+	}
+
+	archives := make([]*archive, 0, len(files))
+
+	// Make sure we don't e.g. see the same filenames in different directories.
+	archiveFilenames := make(map[string]string)
+
+	for _, f := range files {
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
+		// Ignore checksum files: we always generate new ones.
+		if strings.HasSuffix(f, ".sha256") {
+			continue
+		}
+
+		filename := filepath.Base(f)
+		if existingF, ok := archiveFilenames[filepath.Base(f)]; ok {
```

Comment:
This does a case sensitive check, but some OSes are case insensitive (e.g. Windows). We should at least convert the file name to lower case before checking and adding file names to the map.